course5/homework10.2/linked_list.h

```diff
@@ -0,0 +1,148 @@
+		const_iterator(Node const * initial): current(initial) { }
+		const_iterator & operator++() {
+			current = current->next.get();
+			return *this;
+		}
+		const_iterator operator++(int) {
+			const_iterator old = *this;
+			current = current->next.get();
+			return old;
+		}
+		T const & operator*() const { return current->value; }
+		bool operator==(const_iterator const & that) const { return this->current == that.current; }
+		bool operator!=(const_iterator const & that) const { return this->current != that.current; }
+	private:
+		Node const * current{nullptr};
+	};
+
+	LinkedList() = default;
+	LinkedList(LinkedList const & that): LinkedList(that.begin(), that.end()) { }
+	LinkedList(LinkedList && that): head(std::move(that.head)) { }
+
+	template <class InputIt>
+	LinkedList(InputIt const & first, InputIt const & last)
+	{
+		std::unique_ptr<Node> * current{&head};
+		for (auto it = first; it != last; ++it)
+		{
+			*current = std::make_unique<Node>();
+			(*current)->value = *it;
+			current = &((*current)->next);
+		}
+	}
+
+	LinkedList & operator=(LinkedList const & that)
+	{
+		this->head = nullptr;
+		std::unique_ptr<Node> * current{&head};
+		for (auto it = that.begin(); it != that.end(); ++it)
+		{
+			*current = std::make_unique<Node>();
+			(*current)->value = *it;
+			current = &((*current)->next);
+		}
+		return *this;
+	}
+		
+	LinkedList & operator=(LinkedList && that) { this->head = std::move(that.head); return *this; }
+
+	bool operator==(LinkedList const & that) const
+	{
+		auto itl = begin(), itr = that.begin();
+		for (; itl != end() && itr != that.end(); ++itl, ++itr)
+		{
+			if (*itl != *itr)
+				return false;
+		}
+		return itl == end() && itr == that.end();
+	}
+	bool operator!=(LinkedList const & that) const { return !(*this == that); }
+
+	LinkedList::iterator begin() { return iterator(head.get()); }
+	LinkedList::const_iterator begin() const { return const_iterator(head.get()); }
+	LinkedList::iterator end() { return iterator(nullptr); }
+	LinkedList::const_iterator end() const { return const_iterator(nullptr); }
+
+	void push_front(T const & element)
+	{
+		std::unique_ptr<LinkedList<T>::Node> newNode = std::make_unique<LinkedList<T>::Node>();
+		newNode->value = element;
+		newNode->next = std::move(head);
+		head = std::move(newNode);
+	}
+	void push_front(T && element)
+	{
+		std::unique_ptr<LinkedList<T>::Node> newNode = std::make_unique<LinkedList<T>::Node>();
+		newNode->value = std::move(element);
+		newNode->next = std::move(head);
+		head = std::move(newNode);
+	}
+
+	template <class... Args>
+	void emplace_front(Args&& ... args)
+	{
+		std::unique_ptr<LinkedList<T>::Node> newNode = std::make_unique<LinkedList<T>::Node>();
+		newNode->value = T(std::forward<Args>(args)...);
+		newNode->next = std::move(head);
+		head = std::move(newNode);
+	}
+
+	void pop_front() { if (head) head = std::move(head->next); }
```

Comment:
What's the best solution? Document that popping from an empty list is undefined and let the program crash?